3-dynamic-analysis/2-testing/1-go-test.md

```diff
@@ -0,0 +1,261 @@
+			if tc.idx >= 0 {
+				t.Errorf("%s should contain %s at position %d, not %d", tc.text, tc.sub, tc.idx, idx)
+			} else {
+				t.Errorf("%s should not contain %s", tc.text, tc.sub)
+			}
+		}
+	}
+}
+```
+
+Even better, since Go 1.7, we can use subtests, which allow you to have a better control over what
+test cases inside of a test you run. You need to give a name to your test case, which is a good
+practice anyway, and call `t.Run`.
+
+[embedmd]:# (strings_test.go /TestIndex_Subtest/ /^}/)
+```go
+TestIndex_Subtest(t *testing.T) {
+	tt := []struct {
+		name string
+		text string
+		sub  string
+		idx  int
+	}{
+		{"first character", "hello, world", "hello", 0},
+		{"not found", "hello, world", "bye", -1},
+		{"last character", "hello, world", "world", 7},
+	}
+	for _, tc := range tt {
+		t.Run(tc.name, func(t *testing.T) {
+			if idx := strings.Index(tc.text, tc.sub); idx != tc.idx {
+				if tc.idx >= 0 {
+					t.Errorf("%s should contain %s at position %d, not %d", tc.text, tc.sub, tc.idx, idx)
+				} else {
+					t.Errorf("%s should not contain %s", tc.text, tc.sub)
+				}
+			}
+		})
+	}
+}
+```
+
+If you run `go test -v` you'll see the subtest that were ran.
+
+```bash
+$ go test -v
+--- PASS: TestIndex_Subtest (0.00s)
+    --- PASS: TestIndex_Subtest/first_character (0.00s)
+    --- PASS: TestIndex_Subtest/not_found (0.00s)
+    --- PASS: TestIndex_Subtest/last_character (0.00s)
+```
+
+### Exercise: subtests
+
+Make your previous test even better by adding more use cases via
+subtests.
+
+## Tests as examples
+
+When you write tests in Go using the `_test` packages, you
+are somehow showing how to use the package. This is so
+useful that Go provides a way to convert those tests into
+examples.
+
+Examples are functions whose name starts with `Example` and
+do not receive any parameters, nor return any values.
+
+The name of an example contains the information of what the example is demoing. In order to write an example for:
+
+- a function `Foo`, write `ExampleFoo`,
+- a method `Foo` on type `Bar`, write `ExampleFooBar`,
+- a type `Foo`, write `ExampleFoo`
+
+During tests we can verify the results by writing code, in examples
+we have a more limited solution: by checking the expected output
+of executing the example.
+
+To do this, we simply add a comment of the form:
+
+```go
+func ExmampleHelloWorld() {
+    fmt.Println("hello, world")
+    // Output:
+    // hello, world
+}
+```
+
+### Exercise: writing examples
+
+Write an example for `Sum` and see it running on your local `godoc` web
+server.
+Use an `Ouput` comment to make sure that if you run `go test` and the
+output doesn't match the example will fail.
+
+## Congratulations
+
+You're now able to write unit tests and examples, and you're able to run them
+on any package you wish. That's pretty awesome!
+
+Next we're going to learn how to figure out what parts of your code you should
+be testing with [code coverage](2-code-coverage.md).
+
```

Comment:
very good points!

go test -race and the fuzzer are very powerful

I dropped the tracing side of the workshop for now, but I'll add trace, perf, and dtrace in the list of issues